src/widgets/keyboard/lv_keyboard.c

```diff
@@ -371,14 +582,314 @@ void lv_keyboard_def_event_cb(lv_event_t * e)
         }
     }
     else {
+#if LV_USE_KEYBOARD_PINYIN
+        if(keyboard->mode != LV_KEYBOARD_MODE_PINYIN) {
+            lv_textarea_add_text(keyboard->ta, txt);
+        }
+        else {
+            if(*txt < 'a' || *txt > 'z') goto update_candidate;
+
+            if(lv_obj_has_flag(keyboard->candidate_cont, LV_OBJ_FLAG_HIDDEN)) {
+                lv_obj_remove_flag(keyboard->candidate_cont, LV_OBJ_FLAG_HIDDEN);
+            }
+
+            if(lv_obj_has_flag(keyboard->pinyin_label, LV_OBJ_FLAG_HIDDEN)) {
+                lv_obj_remove_flag(keyboard->pinyin_label, LV_OBJ_FLAG_HIDDEN);
+            }
+
+            if(lv_obj_has_flag(keyboard->candidate_btnm, LV_OBJ_FLAG_HIDDEN)) {
+                lv_obj_remove_flag(keyboard->candidate_btnm, LV_OBJ_FLAG_HIDDEN);
+            }
+
+            lv_label_ins_text(keyboard->pinyin_label, LV_LABEL_POS_LAST, txt);
+
+            lv_keyboard_proc_pinyin(keyboard);
+        }
+#else
         lv_textarea_add_text(keyboard->ta, txt);
+#endif
     }
+
+update_candidate:
+#if LV_USE_KEYBOARD_PINYIN
+    lv_keyboard_hidden_candidate(keyboard);
+#else
+    return;
+#endif
 }
 
 /**********************
  *   STATIC FUNCTIONS
  **********************/
 
+#if LV_USE_KEYBOARD_PINYIN
+static void lv_keyboard_free_candidate_btnm_map(lv_keyboard_t * keyboard)
+{
+    if(keyboard->candidate_num != 0) {
+        for(uint16_t i = 0; i < keyboard->candidate_num; i++) {
+            lv_free(keyboard->candidate_btnm_map[i]);
+        }
+        lv_free(keyboard->candidate_btnm_map);
+        keyboard->candidate_num = 0;
+    }
+}
+
+#if LV_KEYBOARD_PINYIN_USE_EXT_BUTTONMATRIX
+static void lv_keyboard_free_ext_candidate_btnm_map(lv_keyboard_t * keyboard)
+{
+    if(keyboard->ext_candidate_num != 0) {
+        for(uint16_t i = 0; i < keyboard->ext_candidate_num; i++) {
+            lv_free(keyboard->ext_candidate_btnm_map[i]);
+        }
+        lv_free(keyboard->ext_candidate_btnm_map);
+        keyboard->ext_candidate_num = 0;
+    }
+}
+#endif
+
+static char * lv_keyboard_search_chinese(size_t * candidate_num, lv_keyboard_t * keyboard, const char * pinyin_str)
+{
+    /*illegal data*/
+    if(*pinyin_str == '\0')    return NULL;
+    if(*pinyin_str == 'i')     return NULL;
+    if(*pinyin_str == 'u')     return NULL;
+    if(*pinyin_str == 'v')     return NULL;
+    if(*pinyin_str == ' ')     return NULL;
+
+    size_t pinyin_str_len = strlen(pinyin_str);
+    char new_pinyin_str[pinyin_str_len + 1];
```

Comment:
Somebody reported this line may not  compile on MSVC. Maybe we can use malloc here? Is `pinyin_str` max length known?

Reply:
> Somebody reported this line may not compile on MSVC. Maybe we can use malloc here? Is `pinyin_str` max length known?

The code does not limit the max length of `pinyin_str`, `pinyin_str` comes from label(`lv_label_get_text()`).